### PR TITLE
Try to find an allocator when the engine is not assigned a device.

### LINKED ISSUE #21487: TensorRT Can't identify the cuda device
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Ubuntu 16.04
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**:
- **TensorFlow installed from (source or binary)**:
source
- **TensorFlow version (use command below)**:
('v1.9.0-rc2-1924-g054b046', '1.10.0-rc1'). Current master branch
- **Python version**: 2.7
- **Bazel version (if compiling from source)**: 0.15.2
- **GCC/Compiler version (if compiling from source)**: 5.4
- **CUDA/cuDNN version**: CUDA 9.0 cuDNN 7.1.4
- **GPU model and memory**: Titan V
- **Exact command to reproduce**:
First train the example small model use [mnist.py](https://gist.github.com/qinyao-he/60171806c40b7b46354234b896c7c2d5).
Then use tensorflow built-in tools to freeze the graph:
```
python -m tensorflow.python.tools.freeze_graph --input_graph log/graph.pbtxt --input_checkpoint log/model.ckpt-20000 --output_node_names softmax_tensor --output_graph log/freeze_graph.pb
```
Finally use [tensorrt.py](https://gist.github.com/qinyao-he/28ddedb7f561bb3cb4ba880833f14a89) to optimize the graph use TensorRT engine.

### Describe the problem
The log shows TensorRT could not find cuda devices. And the graph remains unchanged after the conversion.

### PR DESCRIPTION
**Pull request by aaroey**
This PR will fix #21487.
+cc @pooyadavoodi.

### CODE REVIEW

**Comment by aaroey** [comment]
Friendly ping @pooyadavoodi.

File: tensorflow/contrib/tensorrt/convert/convert_graph.cc [hunk 1]
```diff
@@ -772,33 +775,54 @@ std::pair<int, tensorflow::Allocator*> GetDeviceAndAllocator(
...
+        cuda_device_id = cuda_gpu_id.value();
+        GPUOptions gpu_options;
+        // If the TF to Cuda gpu id mapping exist, the device and corresponding
+        // allocator must have been initialized already, so the
+        // GetGPUAllocator() call won't create a new allocator.
+        dev_allocator = GPUProcessState::singleton()->GetGPUAllocator(
+            gpu_options, tf_gpu_id, 1);
+        break;
       }
+      VLOG(2) << "TF GPU with id " << tf_gpu_id_value << " do not exist " << s;
```
[the remark this answers is not in the document]
Comment (aaroey) [reply tensorflow/contrib/tensorrt/convert/convert_graph.cc]: Done.

File: tensorflow/contrib/tensorrt/convert/convert_graph.cc [hunk 2]
```diff
@@ -772,33 +775,54 @@ std::pair<int, tensorflow::Allocator*> GetDeviceAndAllocator(
...
+      LOG(WARNING) << msg;
     }
+    tensorflow::AllocatorAttributes alloc_attr;
+    cuda_device_id = devices[0]->tensorflow_gpu_device_info()->gpu_id;
+    dev_allocator = devices[0]->GetAllocator(alloc_attr);
+    VLOG(1) << "Using allocator " << dev_allocator->Name()
+            << " and cuda_device_id " << cuda_device_id;
+  } else {
+    LOG(WARNING) << "Cluster is set but device '" << engine.device
+                 << "' is not found in the cluster";
```
[the remark this answers is not in the document]
Comment (aaroey) [reply tensorflow/contrib/tensorrt/convert/convert_graph.cc]: In that case `cuda_device_id` remains -1 and `dev_allocator` is `nullptr`.

File: tensorflow/contrib/tensorrt/convert/convert_nodes.cc [hunk 3]
```diff
@@ -77,6 +77,9 @@ limitations under the License.
 
 namespace tensorflow {
 namespace tensorrt {
+const char* const kInputPHName = "TensorRTInputPH_";
```
[the remark this answers is not in the document]
Comment (aaroey) [reply tensorflow/contrib/tensorrt/convert/convert_nodes.cc]: Of course it can, I just cannot find a good place to put. Added TODO.

**Review by aaroey — COMMENTED** [review]
Thanks for the review, PTAL.

File: tensorflow/contrib/tensorrt/convert/convert_graph.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (pooyadavoodi) [reply tensorflow/contrib/tensorrt/convert/convert_graph.cc]: Okay. Makes sense.

File: tensorflow/contrib/tensorrt/convert/convert_graph.cc [hunk 4]
```diff
@@ -772,33 +775,54 @@ std::pair<int, tensorflow::Allocator*> GetDeviceAndAllocator(
...
-    if (!engine.device.empty() && params.cluster->GetDeviceSet()) {
-      DeviceNameUtils::ParsedName parsed_name;
-      if (DeviceNameUtils::ParseFullName(engine.device, &parsed_name) &&
-          parsed_name.has_id) {
-        params.cluster->GetDeviceSet()->FindMatchingDevices(parsed_name,
-                                                            &devices);
+  if (params.cluster == nullptr || params.cluster->GetDeviceSet() == nullptr ||
+      engine.device.empty()) {
+    // If device is not set, use the first found GPU device for the conversion.
+    for (int tf_gpu_id_value = 0; tf_gpu_id_value < 100; ++tf_gpu_id_value) {
```
[the remark this answers is not in the document]
Comment (pooyadavoodi) [reply tensorflow/contrib/tensorrt/convert/convert_graph.cc]: Got it. Looks okay.

**Comment by pooyadavoodi** [comment]
The PR looks good to me. Thanks @aaroey

**Review by jjsjann123 — APPROVED** [review]
Thanks for the update.
approve per Pooya's review

**Pull request merged**